llama-index-integrations/vector_stores/llama-index-vector-store-firestore/llama_index/vector_store/firestore/base.py

```diff
@@ -0,0 +1,218 @@
+            }
+            ids.append(node_id)
+            entries.append(entry)
+        self._upsert_batch(entries, ids)
+        return ids
+
+    def delete(self, ref_doc_id: str, **delete_kwargs: Any) -> None:
+        """
+        Delete nodes using with ref_doc_id."""
+        self._client.collection(self.collection_name).document(ref_doc_id).delete()
+
+    def query(self, query: VectorStoreQuery, **kwargs: Any) -> VectorStoreQueryResult:
+        """Query vector store."""
+        if query.query_embedding is None:
+            raise ValueError("Query embedding is required.")
+
+        k = kwargs.get("k") or DEFAULT_TOP_K
+        filters = _to_firestore_filter(query.filters) if query.filters else None
+        results = self._similarity_search(
+            query.query_embedding, k, filters=filters, **kwargs
+        )
+
+        top_k_ids = []
+        top_k_nodes = []
+
+        for result in results:
+            result_dict = result.to_dict() or {}
+            if self.metadata_key:
+                metadata = result_dict.get(self.metadata_key)
+
+            node = metadata_dict_to_node(metadata)
+
+            if result_dict.get(self.text_key) is not None:
+                node.set_content(result_dict.get(self.text_key))
+
+            top_k_ids.append(result.id)
+            top_k_nodes.append(node)
+
+        final_result = VectorStoreQueryResult(nodes=top_k_nodes, ids=top_k_ids)
+        _logger.debug("Result of query: %s", final_result)
+        return final_result
+
+    def _upsert_batch(self, entries: List[dict], ids: Optional[List[str]]) -> None:
+        """Upsert batch of vectors to Firestore."""
+        if ids and len(ids) != len(entries):
+            raise ValueError("Length of ids and entries should be the same.")
+
+        db_batch = self._client.batch()
+
+        for batch in more_itertools.chunked(entries, DEFAULT_BATCH_SIZE):
+            for i, entry in enumerate(batch):
+                # Convert the embedding array to a Firestore Vector
+                entry[self.embedding_key] = Vector(entry[self.embedding_key])
+                doc = self._client.collection(self.collection_name).document(
+                    ids[i] if ids else None
+                )
+                db_batch.set(doc, entry, merge=True)
+            db_batch.commit()
+
+    def _similarity_search(
+        self, query: List[float], k: int = DEFAULT_TOP_K, **kwargs: Any
+    ) -> List[DocumentSnapshot]:
+        _filters = kwargs.get("filters")
+        print(_filters)
+        _logger.info("Querying Firestore with filters: %s", _filters)
+
+        wfilters = None
+        collection = self._client.collection(self.collection_name)
+
+        if _filters is not None:
+            wfilters = collection.where(filter=_filters)
+
+        results = (wfilters or collection).find_nearest(
+            vector_field=self.embedding_key,
+            query_vector=Vector(query),
+            distance_measure=self.distance_strategy,
```

Comment:
Should distance_measure be configurable based on args?

Reply:
this was the spec used for LangChain vector store as well